client/error.py

```diff
@@ -240,6 +265,76 @@ def to_text(self) -> str:
         column = click.style(str(self.column), fg="yellow")
         return f"{path}:{line}:{column} {self.description}"
 
+    def to_sarif(self) -> Dict[str, Any]:
+        return {
+            "ruleId": "PYRE-ERROR-" + str(self.code)
+            if self.code is not None
+            else "PYRE-ERROR-MDL",
+            "level": "error",
+            "message": {"text": self.description},
+            "locations": [
+                {
+                    "physicalLocation": {
+                        "artifactLocation": {
+                            "uri": str(self.path),
+                        },
+                        "region": {
+                            "startLine": self.line + 1,
+                            "startColumn": self.column + 1,
+                            "endLine": self.stop_line + 1,
+                            "endColumn": self.stop_column + 1,
+                        },
+                    },
+                },
+            ],
+        }
+
+
+def errors_to_sarif(
+    errors: Union[
+        Sequence[Error], Sequence[LegacyError], Sequence[ModelVerificationError]
+    ]
+) -> Dict[str, Any]:
+
+    results: List[Dict[str, Any]] = []
+    rules: List[Dict[str, Any]] = []
+    for error in errors:
+        results.append(error.to_sarif())
+        if type(error) is LegacyError:
+            error = error.error
+        if type(error) is ModelVerificationError:
+            name = "Model verification error"
+        else:
+            name = error.name
+        rules.append(
+            {
+                "id": "PYRE-ERROR-" + str(error.code)
+                if error.code is not None
+                else "PYRE-ERROR-MDL",
+                "name": name.title().replace(" ", ""),
+                "shortDescription": {"text": name},
+                "helpUri": "https://www.pyre-check.org",
+                "help": {"text": name},
+            }
+        )
```

Comment:
Looking at the context, I think it's more consistent to push these error-specific `rules` computation into individual error class themselves (just like how `results` were computed). E.g. either make `Error`, `LegacyError` and `ModelVerificationError` all have two methods `to_sarif_result` and `to_sarif_rule`, or have their `to_sarif` method return a dataclass that has both a result and a rule field.

Reply:
@grievejia thank you for your time and looking into this. You are right! Spot on - refactored as per suggestion by adding methods to all 3 error classes.